crates/error/src/error.rs

```diff
@@ -0,0 +1,1812 @@
+/// #![cfg(all(feature = "backtrace", not(miri)))]
+/// # let _ = unsafe { std::env::set_var("RUST_BACKTRACE", "1") };
+/// # use wasmtime_internal_error as wasmtime;
+/// use wasmtime::{bail, Context as _, Result};
+///
+/// fn uno() -> Result<()> {
+///     bail!("ouch")
+/// }
+///
+/// fn dos() -> Result<()> {
+///     uno().context("whoops")
+/// }
+///
+/// fn tres() -> Result<()> {
+///     dos().context("uh oh")
+/// }
+///
+/// let error = tres().unwrap_err();
+///
+/// println!("{error}");
+/// // Prints:
+/// //
+/// //     uh oh
+///
+/// println!("{error:#}");
+/// // Prints:
+/// //
+/// //     uh oh: whoops: ouch
+///
+/// println!("{error:?}");
+/// // Prints
+/// //
+/// //     uh oh
+/// //
+/// //     Caused by:
+/// //         0: whoops
+/// //         1: ouch
+/// //
+/// //     Stack backtrace:
+/// //       <...>
+/// //        7: example::uno
+/// //        8: example::dos
+/// //        9: example::tres
+/// //       10: example::main
+/// //       <...>
+///
+/// println!("{error:#?}");
+/// // Prints
+/// //
+/// //     Error {
+/// //         <...>
+/// //     }
+/// # }
+/// ```
+///
+/// # Converting a `wasmtime::Error` into an `anyhow::Error`
+///
+/// When the `"anyhow"` feature is enabled, there is a `From<wasmtime::Error>
+/// for anyhow::Error` implementation. You can always call that implementation
+/// explicitly if needed, but `?`-propagation allows the conversion to happen
+/// seamlessly from functions that return a `Result<T, wasmtime::Error>` to
+/// those that return a `Result<U, anyhow::Error>`.
+///
+/// ```
+/// # fn _foo() {
+/// #![cfg(feature = "anyhow")]
+/// # use wasmtime_internal_error as wasmtime;
+///
+/// fn foo() -> Result<(), wasmtime::Error> {
+///     wasmtime::bail!("decontamination failure")
+/// }
+///
+/// fn bar() -> Result<(), anyhow::Error> {
+///     foo()?; // `?` is auto-converting here!
+///     Ok(())
+/// }
+///
+/// let error = bar().unwrap_err();
+/// assert_eq!(error.to_string(), "decontamination failure");
+/// # }
+/// ```
+///
+/// # Converting an `anyhow::Error` into a `wasmtime::Error`
+///
+/// When the `"anyhow"` feature is enabled, there is an `Error::from_anyhow`
+/// constructor that you may use to convert an `anyhow::Error` into a
+/// `wasmtime::Error`. (Unfortunately trait coherence does not allow us a
+/// `From<anyhow::Error> for wasmtime::Error` implementation.) This will
+/// most-often be used in combination with `Result::map_err`:
+///
+/// ```
+/// # fn _foo() {
+/// #![cfg(feature = "anyhow")]
+/// # use wasmtime_internal_error as wasmtime;
+///
+/// fn baz() -> Result<(), anyhow::Error> {
+///     anyhow::bail!("oops I ate worms")
+/// }
+///
+/// fn qux() -> Result<(), wasmtime::Error> {
+///     baz().map_err(wasmtime::Error::from_anyhow)?;
+///     Ok(())
+/// }
+///
+/// let error = qux().unwrap_err();
+/// assert_eq!(error.to_string(), "oops I ate worms");
+/// # }
+/// ```
+pub struct Error {
+    pub(crate) inner: OomOrDynError,
+}
```

Comment:
This might also benefit from `#[repr(transparent)]` to ensure that ABI considerations don't think that this is a struct